Indicate goodness of fit of regression line

atdotde opened the pull request:
The goodness of fit of a regression line is the percentage
of the variance of the y values that is explained by the
dependence on the x values.

Set the alpha value of the regression line to this goodness
of fit.

Further, set the width of the regression line to a standard
deviation of the values from the regression line valies.

Signed-off-by: Robert C. Helling <helling@atdotde.de>

<!-- Lines like this one are comments and will not be shown in the final output. -->
<!-- Make sure that you have read the "Contributing" section of the README and also the notes in CodingStyle. -->
<!-- If you are a collaborator, please add labels and assign other collaborators for a review. -->

### Describe the pull request:
<!-- Replace [ ] with [x] to select options. -->
- [ ] Bug fix
- [ ] Functional change
- [x] New feature
- [ ] Code cleanup
- [ ] Build system change
- [ ] Documentation change
- [ ] Language translation

### Pull request long description:
<!-- Describe your pull request in detail. -->

### Changes made:
<!-- Enumerate the changes with 1), 2), 3) etc. -->
<!-- Ensure the test cases are updated if needed. -->

### Related issues:
<!-- Reference issues with #<issue-num>. -->
<!-- Write "Fixes #<issue-num" to notify Github that this PR fixes an issue. -->

### Additional information:
<!-- Include sample dive log or other relevant information to allow testing the change where feasible. -->

### Release note:
<!-- Describe if this change needs a release note present in CHANGELOG.md. -->
<!-- Also, please make sure to add the release note on top of the file CHANGELOG.md. -->

### Documentation change:
<!-- If this PR makes changes to user functionality, then the documentation has to be updated too. -->
<!-- Please, briefly outline here what has changed in terms of the user experience (UX). -->
<!-- If UX changes have been made, a maintainer should apply the 'needs-documentation-change' label. -->

### Mentions:
<!-- Mention users that you want to review your pull request with @<user-name>. Leave empty if not sure. -->

dirkhh commented:
Ugh - screenshots that explain what that does? :-)

bstoeger commented on stats/statsview.cpp:
```diff
@@ -1023,10 +1030,10 @@ static bool is_linear_regression(int sample_size, double cov, double sx2, double
 // Returns the coefficients [a,b] of the line y = ax + b
 // If case of an undetermined regression or one with infinite slope, returns [nan, nan]
-static std::pair<double, double> linear_regression(const std::vector<StatsScatterItem> &v)
+static std::tuple<double, double, double, double> linear_regression(const std::vector<StatsScatterItem> &v)
```
I have no time to look closer today, though from a cursory glance it looks good.

Just one comment: here I would suggest returning a struct for documentation purposes. The structured binding below will work anyway. Also the comment above is of course not correct anymore.

bstoeger commented:
I would also like to ask @willemferguson to comment, as the statistical significance code is from him.

dirkhh commented:
Visually I'd like to see the center of the line drawn as it used to be; overlaying the standard deviation with a color indicating the goodness is a nice addition - having a label 'somewhere' that provides an explanation / value of the goodness is almost required for this to be meaningful. Otherwise it's just very confusing.

atdotde commented:
<img width="1202" alt="Screenshot 2021-01-10 at 11 58 26" src="https://user-images.githubusercontent.com/863957/104120983-445c2f00-533b-11eb-8c4e-4d3c0f792db2.png">
<img width="1214" alt="Screenshot 2021-01-10 at 11 58 50" src="https://user-images.githubusercontent.com/863957/104120985-47571f80-533b-11eb-9603-ad4f57a8a0c2.png">

I could add the original central line. But the point is that central line is, strictly speaking, not really known due to statistical fluctuations. The area that is now coloured is where likely (within one standard deviation) the "true" regression line would be, the data suggests that our knowledge is only as precise as the width of that line.

bstoeger commented on stats/statsview.cpp:
```diff
@@ -1021,12 +1028,21 @@ static bool is_linear_regression(int sample_size, double cov, double sx2, double
 {
 	if (v.size() < 2)
-		return { NaN, NaN };
+		return { .a = NaN, .b = NaN, .res2 = 0.0, .r2 = 0.0};
```
Note that this syntax is C++20 only. Somewhat surprisingly, it appears to work with our compilers.

bstoeger commented on stats/statsview.cpp:
```diff
@@ -1080,10 +1101,10 @@ void StatsView::plotScatter(const std::vector<dive *> &dives, const StatsVariabl
 		auto [minx, maxx] = axisX->minMax();
 		auto [miny, maxy] = axisY->minMax();
-		addLinearRegression(a, b, minx, maxx, miny, maxy, xAxis, yAxis);
+		addLinearRegression(reg.a, reg.b, reg.res2, reg.r2, minx, maxx, miny, maxy, xAxis, yAxis);
```
Why not pass the structure instead of the four members separately?

If you don't want to export the structure in the ".h" file, you can simply forward declare it and pass a const reference.

bstoeger commented on stats/statsview.cpp:
```diff
@@ -747,6 +747,9 @@ void StatsView::RegressionLine::updatePosition()
 		maxX = std::min(maxX, intersect_x2);
 	}
+	QPen pen = item->pen();
+	pen.setWidthF(-2.0 * width * (yAxis->toScreen(maxY)-yAxis->toScreen(minY)) / (maxY - minY));
```
I'm not very fond of the optics of the screen shot that you posted. I think the line shouldn't cross the axes. Therefore, I suggest doing this with a QPolygonF.

You can clip the polygon with QPolygonF::intersected() function to the plotting area.

That said, is the pen width really the right thing to do here? Note that the width is perpendicular to the line direction, but naively I would assume that the width you want to specify is perpendicular to the y-axis...?

atdotde commented:
I will try to do this with the polygon. 

Regarding the width, I have to admit, I thought about this (the width being measured orthogonal to the line while the standard deviation being in the y direction) but ended up being too lazy to do it. But spending two more minutes of though and a sketch shows that the width should be multiplied by sqrt(1+a^2) to correct for this. Fixing this should be easy.

willemferguson commented:
I need to repeat:

Given a non-significant regression relationship between two variables, 
indicating any regression line at all is meaningless. In fact it is 
actively misleading, making you think that there may be a relationship. 
The slope of such a line has no meaning and henceforth is not useful for 
any interpretation.

The broad regression area that Robert indicates in his two diagrams are 
also not useful at all. If one were to plot a pink area of reliability, 
then a 1 standard deviation line is also not useful because it does not 
allow any reliable interpretation of the slope. If one were to do 
anything like this, one needs to indicate the 95% reliability area 
around the regression line, not a 1-sd line. The 1-sd line does not 
correspond to the statistical calculation that was done to generate the 
line. And then only individuals with some background in statistics would 
be able to interpret the 95% reliability area.

Honestly, my recommendation is not to show any regression line if there 
is not a statistically significant trend.

However, Dirk is correct in that it would help very much is some 
information about the regression line is shown. Probably something like 
"Linear regression line". If users could tolerate the statistical 
jargon, one could add the statistical probability, e.g. "p < 0.01".

I hope this makes sense???

Kind regards,

willem



-- 
This message and attachments are subject to a disclaimer.

Please refer to 
http://upnet.up.ac.za/services/it/documentation/docs/004167.pdf 
<http://upnet.up.ac.za/services/it/documentation/docs/004167.pdf> for
full 
details.

willemferguson commented:
On 2021/01/10 14:32, Robert C. Helling wrote:
>
> I will try to do this with the polygon.
>
> Regarding the width, I have to admit, I thought about this (the width 
> being measured orthogonal to the line while the standard deviation 
> being in the y direction) but ended up being too lazy to do it. But 
> spending two more minutes of though and a sketch shows that the width 
> should be multiplied by sqrt(1+a^2) to correct for this. Fixing this 
> should be easy.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub 
> <https://github.com/subsurface/subsurface/pull/3148#issuecomment-757468696>, 
> or unsubscribe 
> <https://github.com/notifications/unsubscribe-auth/ABGX5MTVX36D5SHF7Q7OKVLSZGM7DANCNFSM4V36HGMQ>.
>
My honest opinion? Do not waste your time time to do this. As indicated 
elsewhere, the statistical validity of such an approach is questionable.


-- 
This message and attachments are subject to a disclaimer.

Please refer to 
http://upnet.up.ac.za/services/it/documentation/docs/004167.pdf 
<http://upnet.up.ac.za/services/it/documentation/docs/004167.pdf> for
full 
details.

atdotde commented:
Here is the version with a polygon that automatically also fixes the issue width vs height in y direction.
<img width="1206" alt="Screenshot 2021-01-10 at 22 40 50" src="https://user-images.githubusercontent.com/863957/104136280-6a162200-5395-11eb-8974-adb45323bda0.png">
<img width="1194" alt="Screenshot 2021-01-10 at 22 41 04" src="https://user-images.githubusercontent.com/863957/104136283-6da9a900-5395-11eb-83fb-460a50ea0eb5.png">

Regarding @willemferguson 's objections: I did not change the check that decides if to draw a line at all (based on the t-test regarding the null hypothesis "slope 0").

But I think I am doing the right thing re "confidence interval": If we assume (which according to https://en.wikipedia.org/wiki/Simple_linear_regression#Normality_assumption is the standard assumption) that the y variable is a function of x and some other factors that we don't directly observe and the dependence on x is affine (linear with intercept) than we can model y as ax+b+residual where residual in normal distributed with mean 0. Then we can calculate the variance of the residual from the data points. This is the best guess for the standard deviation of the residuals.

So, now, if the (secret) true dependence is y = a_true * x + b_true + normal_distribution(of said variance) and we draw N points. From those points, we are computing our guess for a and b (which will fluctuate around the true values a_true and b_true). The question is by how much. Then my claim (supported by the cited wikipedia article) is that within one standard deviation, i.e. with confidence interval 68% the true regression line will be inside the band I am plotting (if you want 95% simply take two standard deviations, i.e. double the width of the band).

So yes, our best guess for the true line is the central line but it could be any line within the band with confidence 68%. This is why I think it is more honest to draw this band rather than just the central line.

bstoeger commented on stats/statsview.cpp:
```diff
@@ -719,13 +719,17 @@ void StatsView::QuartileMarker::updatePosition()
+	QColor invisible(Qt::white);
+	invisible.setAlpha(0);
+	QPen pen(invisible);
 	item->setPen(pen);
```
I think you can replace the previous four lines by
```
item->setPen(Qt::NoPen);
```
At the very least you can replace the color thing by `Qt::transparent`.

bstoeger commented on stats/statsview.cpp:
```diff
@@ -719,13 +719,13 @@ void StatsView::QuartileMarker::updatePosition()
-	a(a), b(b)
+	a(a), b(b), width(width)
 {
-	item->setZValue(ZValues::chartFeatures);
```
Sorry - I think now you removed one line too much: The zValue still has to be set.

If that is fixed, consider it approved by me. I can resolve the conflict, if you want.

atdotde replied:
You are right. I had experimented with the z value to make the dots stand out more but that was not effective. So I wanted to remove that again and accidentally removed the whole line.

willemferguson commented:
The problem with a confidence interval for a regression line is that you should not only take into account the y-axis intersect (as you do), but you also need to take into account the error around the estimate of the slope parameter. Therefore the confidence interval has a concave shape, not a parallelogram.

atdotde commented:
Rebased on master and added a commit with suggestions from clang.

@willemferguson Ah, I think the misunderstanding is what the broad bar is supposed to mean: It is not supposed to indicate that the regression line could be shifted up or down (as it would come from varying the y intercept). But it should indicate all straight lines within the bar (including those with different slope). As described in the wikipedia article quoted above, the uncertainty (standard deviation) of the slope is given by the square root of the variance of the residuals (which are the width that I use) divided by the standard deviation of the x values. But since our box is essentially as wide as the spreading of the x values, all lines that you get by varying the slope in that interval end up (essentially) in that bar (as long as you let them go though the center of mass of the dots).

atdotde commented:
I mean, these would all be lines compatible with the region in slope and intercept within one standard deviation.
![Screenshot 2021-01-10 at 22 40 50](https://user-images.githubusercontent.com/863957/104183592-7c827100-5412-11eb-870f-d17c4ab025b8.jpeg)

bstoeger reviewed (approved):
The code is fine with me - I cannot comment on the statistical things.

willemferguson commented:
If I understand you correctly, your Internet source is plain misleading. Below is a conceptual drawing of the sources of error and of the confidence region of a linear regression model. The regression line has a reference in the middle of the diagram, conceptually located at (mean-of-Yvalues, mean-of-Xvalues). The regression line can rotate around this point BUT ALSO, as you note, move up and down within the confidence region. Note that the confidence region is defined by the critical p-value used in the analysis, and not by the residuals of the y-values. I suggest, for software aimed at laypeople, we do not touch this topic. In order to really interpret the confidence area, one needs to understand the characteristics of the sums-of-squares of which the regression is composed. If one want to move into this area, one uses professional statistical software.

![reg1](https://user-images.githubusercontent.com/5078706/104196263-3f2add00-542c-11eb-84a3-36fe93ba1c52.jpg)

atdotde commented:
I agree with your drawing. But since the window on the x axis that we show is given by the range of x values, it's size is comparable to the spread of the x values which is exactly the range over which your lines bend, i.e. it is the range where straight boundaries are a good approximation. This after all is mainly an order of magnitude visual guide telling the user that the true regression line is likely somewhere in that region and it typically off by an amount indicated by the width. 

As these are diving related quantities they will depend on a million of factors (my plot examples where mean depth as a function of time and SAC as a function of time) so I wanted to indicate by how much that regression line is to be trusted for which an order of magnitude estimate should be good enough (and for which the assumption of normally distributed residuals which allows the errors to be estimated using the variance of the residuals should be good enough as well). 

I guess all I am saying is that the regression line is not as precise as a thin line indicates but has statistical uncertainty for which the shaded region gives an idea about how much to expect. Thin region: varying x explains varying y pretty well, thick region: there are strong other factors.

willemferguson commented:
Please explain a bit more completely? Here is  a graph of a 95% confidence area created using R.  As you can see, in this case, the linear model fits the data pretty well. The curved areas cover the whole range of x. I cannot see where a linear approximation would suffice. The reason is that the error in the slope estimate is not accounted for by a linear approximation.
![Reg2](https://user-images.githubusercontent.com/5078706/104284660-084fd800-54bb-11eb-9dbb-a1e82d8da539.png)

dirkhh commented:
I haven't been using my math degree in 25 years, so I may be a bit rusty... but what @willemferguson writes makes a lot of sense to me. We have uncertainty as to where that central point might be, and we have uncertainty about the slope. To me it seems logical that that would create this concave shape that is shaded in that last picture. I have a much easier time wrapping my mind around that then around the parallelogram that @atdotde 's code creates. I really like the visualization suggested with the central line drawn and then the area around it shaded.

Robert?

bstoeger commented:
Could it be that one is the "confidence interval" and the other the "prediction interval"?

If it is only a naming issue, I suggest implementing this as a "chart feature", as was requested on the mailing list. Thereby it is automatically given a user visible name.

atdotde commented:
OK, let's do the right thing and plot the concave confidence intervals.
<img width="1229" alt="Screenshot 2021-01-12 at 19 44 24" src="https://user-images.githubusercontent.com/863957/104358330-9608e300-550e-11eb-8cba-3549010818a3.png">
Previously, I had also missed a factor of 1/sqrt(n) in the residuals.
<img width="1228" alt="Screenshot 2021-01-12 at 19 45 18" src="https://user-images.githubusercontent.com/863957/104358420-b638a200-550e-11eb-9b77-6b49ea89989b.png">

dirkhh commented:
I know we are approaching 'bike-shedding' - but any chance you could plot the center line in a slightly darker color?

@willemferguson  -- what do you think about this latest version? Still vehemently against it?

atdotde commented:
@dirkhh I understand you don't want to leave your bicycle in the rain. I hope you like your shed in Qt::red.

<img width="1230" alt="Screenshot 2021-01-12 at 21 26 32" src="https://user-images.githubusercontent.com/863957/104369035-2f3ef600-551d-11eb-8bf5-7d734a2f37c5.png">

Honestly, looking at this, I feel that the job of the anti-aliasing could be done better as the line has visible stairs. I tried a few things (including removing the confidence region altogether) but none helped.

dirkhh commented:
> @dirkhh I understand you don't want to leave your bicycle in the rain. I hope you like your shed in Qt::red.

That is so kind of you!
 
> Honestly, looking at this, I feel that the job of the anti-aliasing could be done better as the line has visible stairs. I tried a few things (including removing the confidence region altogether) but none helped.

I was looking at that and thinking 'what's going on with that line'? That reminds me of X Terminals ca. 1990...

bstoeger commented:
I wouldn't worry about antialiasing when we will replace the line by a `QSGGeometryNode` in due course anyway.

dirkhh commented:
@willemferguson -- any comments / thoughts?

willemferguson commented:
I am easy with what Robert has done. I have not seen this in action, but it appears GTM. A few matters:
1) It has previously been suggested that showing the regression line and the associated confidence area might be a configurable option. I think this would be a good idea.
2) If the regression line and the pink area are shown, the appropriate information needs to be provided. a) Text parallel to the regression line should indicate "Regression line" or similar. I think indicating the regression coefficients for this line is outside of the scope of software such as Subsurface. b) The probability percentage of the confidence area needs to be shown. Conventionally, the 95% confidence area (i.e. with 2.5% of the total probability falling above the confidence area and another 2.5% falling below the area) is shown in such graphs.
Interpreting Robert's graphs above, one notices that drawing a straight line from  the upper lefthand corner of the pink area to the lower righthand corner results in a positive slope (see the blue lines in my mockup below), reflecting the high statistical probability of a relationship between the variables on the horizontal and vertical axes. When there is not a statistically significant relationship between these two variables, a horizontal blue line would fit inside the confidence area. The important point is, however, that the confidence area (+ the possibility of fitting a horizontal line inside) is not the tool for detecting the statistical significance of a regression: the confidence area is constructed post hoc after doing the analysis that determines the statistical probability of a relationship between the two variable. The confidence area is the reflection and consequence of the primary regression analysis that has been done elsewhere and which determine the shape of the pink area. If the primary analysis shows that the regression is not statistically significant, the regression line and confidence area would normally not be shown. I do NOT, for a moment, suggest that my blue lines above be included in the code. I am attempting to explain how one can use the confidence area to make conclusions about the strength of the relationship between the two variables being analysed.
![Screenshot from 2021-01-14 06-40-25](https://user-images.githubusercontent.com/5078706/104546757-93ed7400-5635-11eb-95d7-a67a8b254ca8.png)

atdotde commented:
I take this as a yes. 

Even though I don't think too much text (which would anyway be the same for all scatter plots) would help the diagram. In other places in Subsurface (for example the heat map or the little diagram in the infobox) we don't add explanatory labels but rather explain these in the documentation.

Looking back at the discussion I think the main difference between @willemferguson 's approach and what I was trying to argue for is that Willem comes from the hypothesis testing direction, something like "there is an influence of the x variable on the y variable" for which a statistical test gives a binary yes/no answer (with some prescribed confidence level, say 95%). This is for example used for the question if there should be a regression line at all (only for a significant influence) whereas I was out for a more qualitative understanding of getting a feeling how rigid is the line we are drawing (for which I would be happy to draw something that would allow for both positive or negative correlation as long as I am aware of that and could still take away for example that the correlation is more likely to be positive than negative even though the null result is still possible with a probability greater than 5%). What I am trying to say is, being a physicist who does almost everything in first oder perturbation theory, I was looking more for a continuous answer rather than a yes/no answer. And the old region I was drawing (minus the sqrt(n) error I made of course) was not to be interpreted as yes/no (within 95% etc etc) rather to give the user an idea about the typical size of fluctuations to expect, so as an order of magnitude statement. But the more precise thing is of course more precise and thus better. But still, I think, it would be even better not as a reddish area (again a yes/no thing) but with varying intensity/alpha channel according to a gaussian in the y direction (with the corresponding mu and sigma). But that would add bells to the already existing bicycle shed....

bstoeger commented:
There is no need to add a label, if this is made an optional chart feature. I will do this once I'm done with the QSG* conversion.

The chart feature list will look something like

- [x] linear regression
- [x] 2σ confidence area

Which will make it very clear what the line / area represents.
